vectordatasource/transform.py

```diff
@@ -2976,27 +2988,281 @@ def _merge_lines(linestring_shapes):
+    result = _drop_small_outers_multi(result, area_tolerance)
+
+    if result.is_empty:
+        return []
+
+    return [result]
+
+
+def _union_bounds(a, b):
+    """
+    Union two (minx, miny, maxx, maxy) tuples of bounds, returning a tuple
+    which covers both inputs.
+    """
+
+    if a is None:
+        return b
+    elif b is None:
+        return a
+    else:
+        aminx, aminy, amaxx, amaxy = a
+        bminx, bminy, bmaxx, bmaxy = b
+        return (min(aminx, bminx), min(aminy, bminy),
+                max(amaxx, bmaxx), max(amaxy, bmaxy))
+
+
+# RecursiveMerger is a set of functions to merge geometry recursively in a
+# quad tree.
+#
+# It consists of three functions, any of which can be `id` for a no-op, and
+# all of which take a single argument which will be a list of shapes, and
+# should return a list of shapes as output.
+#
+#   * leaf: called at the leaves of the quad tree with original geometry.
+#   * node: called at internal nodes of the quad tree with the results of
+#           either calls to leaf() or node().
+#   * root: called once at the root with the results of the top node (or leaf
+#           if it's a degenerate single-level tree).
+#
+# These allow us to merge transformed versions of geometry, where leaf()
+# transforms the geometry to some other form (e.g: buffered for buildings),
+# node merges those recursively, and then root reverses the buffering.
+#
+RecursiveMerger = namedtuple('RecursiveMerger', 'leaf node root')
+
+
+# A bucket used to sort shapes into the next level of the quad tree.
+Bucket = namedtuple("Bucket", "bounds box shapes")
+
+
+def _mkbucket(*bounds):
+    """
+    Convenience method to make a bucket from a tuple of bounds (minx, miny,
+    maxx, maxy) and also make the Shapely shape for that.
+    """
+
+    from shapely.geometry import box
+
+    return Bucket(bounds, box(*bounds), [])
+
+
+def _merge_shapes_recursively(shapes, shapes_per_merge, merger, depth=0,
+                              bounds=None):
+    """
+    Group the shapes geographically, returning a list of shapes. The merger,
+    which must be a RecursiveMerger, controls how the shapes are merged.
+
+    This is to help merging/unioning, where it's better to try and merge shapes
+    which are adjacent or near each other, rather than just taking a slice of
+    a list of shapes which might be in any order.
+
+    The shapes_per_merge controls at what depth the tree starts merging.
+    Smaller values mean a deeper tree, which might increase performance if
+    merging large numbers of items at once is slow.
+
+    This method is recursive, and will bottom out after 5 levels deep, which
+    might mean that sometimes more than shapes_per_merge items are merged at
+    once.
+    """
+
+    assert isinstance(merger, RecursiveMerger)
+
+    # don't keep recursing. if we haven't been able to get to a smaller number
+    # of shaped by 5 levels down, then perhaps there are particularly large
```

Comment:
`# of shapes`

Reply:
Good catch of typo

Reply:
Yup, well spotted, thanks! Fixed in 937023cb.